### PR TITLE
Log parser time mean

### PR DESCRIPTION
**Pull request by peterfpeterson**
While walking through code I discovered that LogParser had a parallel implementation of calculating time average mean values. This merges it with the central version in TimeSeriesProperty.

**To test:**

The existing tests should all still pass.

Refs #34794.

*This does not require release notes* because it is a refactor to remove duplicate implementations.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/inc/MantidKernel/ITimeSeriesProperty.h [hunk 1]
```diff
@@ -43,13 +43,11 @@ class ITimeSeriesProperty {
   virtual std::pair<double, double> averageAndStdDevInFilter(const std::vector<SplittingInterval> &filter) const = 0;
   /// Return the time series's times as a vector<DateAndTime>
   virtual std::vector<Types::Core::DateAndTime> timesAsVector() const = 0;
-  /// Returns the calculated time weighted average value
-  virtual double timeAverageValue() const = 0;
```
Comment (jmborr) [inline Framework/Kernel/inc/MantidKernel/ITimeSeriesProperty.h:47]: you'll have to update the exposure of `timeAverageValue` to the python API in `Framework/PythonInterface/mantid/kernel/src/Exports/TimeSeriesProperty.cpp`

**Review by jmborr — CHANGES_REQUESTED** [review]

File: Framework/Kernel/inc/MantidKernel/ITimeSeriesProperty.h (hunk 1 above)
Comment (peterfpeterson) [reply Framework/Kernel/inc/MantidKernel/ITimeSeriesProperty.h:47]: I bet that is a couple of the broken tests right there. Thanks for the catch!

**Review by jmborr — CHANGES_REQUESTED** [review]

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 2]
```diff
@@ -829,15 +819,18 @@ template <typename TYPE>
...
-  // First of all, if the log or the filter is empty, return NaN
-  if (realSize() == 0 || filter.empty()) {
-    return std::numeric_limits<double>::quiet_NaN();
-  }
-
   // If there's just a single value in the log, return that.
   if (realSize() == 1) {
     return static_cast<double>(m_values.front().value());
   }
+  if (size() == 1) {
```
[the remark this answers is not in the document]
Comment (peterfpeterson) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:826]: It is a little more subtle than that. `realSize()` is the length of the vector and `size()` is the number of values with unique times. I haven't verified that it is enforced, but `size() <= realSize()` is an underlying assumption.

[the remark this answers is not in the document]
Comment (jmborr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:826]: `size() <= realSize()` is my point. Condition `if (size() == 1)` makes `if (realSize() == 1)` superfluous.

**Review by jmborr — DISMISSED** [review]

**Review by jmborr — APPROVED** [review]

**Pull request merged**